File: android/app/src/main/java/deckers/thibault/aves/channel/calls/AppShortcutHandler.java
package deckers.thibault.aves.channel.calls;

import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.text.TextUtils;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.pm.ShortcutInfoCompat;
import androidx.core.content.pm.ShortcutManagerCompat;
import androidx.core.graphics.drawable.IconCompat;

import java.util.List;

import deckers.thibault.aves.MainActivity;
import deckers.thibault.aves.R;
import deckers.thibault.aves.utils.BitmapUtils;
import io.flutter.plugin.common.MethodCall;
import io.flutter.plugin.common.MethodChannel;

public class AppShortcutHandler implements MethodChannel.MethodCallHandler {
    public static final String CHANNEL = "deckers.thibault/aves/shortcut";

    private Context context;

    public AppShortcutHandler(Context context) {
        this.context = context;
    }

    @Override
    public void onMethodCall(@NonNull MethodCall call, @NonNull MethodChannel.Result result) {
        switch (call.method) {
            case "canPin": {
                result.success(ShortcutManagerCompat.isRequestPinShortcutSupported(context));
                break;
            }
            case "pin": {
                String label = call.argument("label");
                byte[] iconBytes = call.argument("iconBytes");
                List<String> filters = call.argument("filters");
                new Thread(() -> pin(label, iconBytes, filters)).start();
                result.success(null);
                break;
            }
            default:
                result.notImplemented();
                break;
        }
    }

    private void pin(String label, byte[] iconBytes, @Nullable List<String> filters) {
        if (!ShortcutManagerCompat.isRequestPinShortcutSupported(context) || filters == null) {
            return;
        }

        IconCompat icon = null;
        if (iconBytes != null && iconBytes.length > 0) {
            Bitmap bitmap = BitmapFactory.decodeByteArray(iconBytes, 0, iconBytes.length);
            bitmap = BitmapUtils.centerSquareCrop(context, bitmap, 256);
            if (bitmap != null) {
                icon = IconCompat.createWithBitmap(bitmap);
            }
        }
        if (icon == null) {
            icon = IconCompat.createWithResource(context, R.mipmap.ic_shortcut_collection);
        }

        Intent intent = new Intent(Intent.ACTION_MAIN, null, context, MainActivity.class)
                .putExtra("page", "/collection")
                .putExtra("filters", filters.toArray(new String[0]));

        ShortcutInfoCompat shortcut = new ShortcutInfoCompat.Builder(context, "collection-" + TextUtils.join("-", filters))
                .setShortLabel(label)
                .setIcon(icon)
                .setIntent(intent)
                .build();

        ShortcutManagerCompat.requestPinShortcut(context, shortcut, null);
    }
}

File: android/app/src/main/kotlin/deckers/thibault/aves/channel/calls/AppShortcutHandler.kt
package deckers.thibault.aves.channel.calls

import android.content.Context
import android.content.Intent
import android.graphics.BitmapFactory
import androidx.core.content.pm.ShortcutInfoCompat
import androidx.core.content.pm.ShortcutManagerCompat
import androidx.core.graphics.drawable.IconCompat
import deckers.thibault.aves.MainActivity
import deckers.thibault.aves.R
import deckers.thibault.aves.channel.calls.Coresult.Companion.safe
import deckers.thibault.aves.utils.BitmapUtils.centerSquareCrop
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.GlobalScope
import kotlinx.coroutines.launch
import java.util.*

class AppShortcutHandler(private val context: Context) : MethodCallHandler {
    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "canPin" -> safe(call, result, ::canPin)
            "pin" -> GlobalScope.launch(Dispatchers.IO) { safe(call, result, ::pin) }
            else -> result.notImplemented()
        }
    }

    private fun isSupported() = ShortcutManagerCompat.isRequestPinShortcutSupported(context)

    private fun canPin(@Suppress("UNUSED_PARAMETER") call: MethodCall, result: MethodChannel.Result) {
        result.success(isSupported())
    }

    private fun pin(call: MethodCall, result: MethodChannel.Result) {
        val label = call.argument<String>("label")
        val iconBytes = call.argument<ByteArray>("iconBytes")
        val filters = call.argument<List<String>>("filters")
        if (label == null || filters == null) {
            result.error("pin-args", "failed because of missing arguments", null)
            return
        }

        if (!isSupported()) {
            result.error("pin-unsupported", "failed because the launcher does not support pinning shortcuts", null)
            return
        }

        var icon: IconCompat? = null
        if (iconBytes?.isNotEmpty() == true) {
            var bitmap = BitmapFactory.decodeByteArray(iconBytes, 0, iconBytes.size)
            bitmap = centerSquareCrop(context, bitmap, 256)
            if (bitmap != null) {
                icon = IconCompat.createWithBitmap(bitmap)
            }
        }
        if (icon == null) {
            icon = IconCompat.createWithResource(context, R.mipmap.ic_shortcut_collection)
        }

        val intent = Intent(Intent.ACTION_MAIN, null, context, MainActivity::class.java)
            .putExtra("page", "/collection")
            .putExtra("filters", filters.toTypedArray())

        // multiple shortcuts sharing the same ID cannot be created with different labels or icons
        // so we provide a unique ID for each one, and let the user manage duplicates (i.e. same filter set), if any
        val shortcut = ShortcutInfoCompat.Builder(context, UUID.randomUUID().toString())
            .setShortLabel(label)
            .setIcon(icon)
            .setIntent(intent)
            .build()
        ShortcutManagerCompat.requestPinShortcut(context, shortcut, null)

        result.success(true)
    }

    companion object {
        const val CHANNEL = "deckers.thibault/aves/shortcut"
    }
}